### PR TITLE
test: reproduce behavior of getExecutableDeclaration discussed in #3245

### PR DESCRIPTION
**Pull request by santos-samuel**
Replicating issue #3245 with a test case.

### CODE REVIEW

File: src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/ChangePasswordValidator.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package pt.ist.socialsoftware.edition.ldod.validator;
```
Comment (stone-wall) [inline src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/ChangePasswordValidator.java]: This package name is incorrect, it should be declared as `package spoon.getExecutableDeclarationBugTest.testClasses;`

File: src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/ChangePasswordValidator.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.crypto.password.PasswordEncoder;
```
Comment (stone-wall) [inline src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/ChangePasswordValidator.java]: The spoon project doesn't currently have any dependency on the Spring Framework so these imports/class dependencies will always cause the test to fail.

File: src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/ChangePasswordValidator.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.validation.Errors;
+import org.springframework.validation.Validator;
+
+import pt.ist.socialsoftware.edition.ldod.domain.LdoD;
+import pt.ist.socialsoftware.edition.ldod.domain.LdoDUser;
+import pt.ist.socialsoftware.edition.ldod.forms.ChangePasswordForm;
```
Comment (stone-wall) [inline src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/ChangePasswordValidator.java]: Same as the above. Try to re-create the test case without any extra dependencies on external projects or files.

File: src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/UserController.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package pt.ist.socialsoftware.edition.ldod.controller;
```
Comment (stone-wall) [inline src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/UserController.java]: Should be 
`spoon.getExecutableDeclarationBugTest.testClasses;`

File: src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/UserController.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Controller;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import pt.ist.socialsoftware.edition.ldod.domain.LdoD;
+import pt.ist.socialsoftware.edition.ldod.domain.LdoDUser;
+import pt.ist.socialsoftware.edition.ldod.forms.ChangePasswordForm;
+import pt.ist.socialsoftware.edition.ldod.validator.ChangePasswordValidator;
```
Comment (stone-wall) [inline src/test/java/spoon/getExecutableDeclarationBugTest/testClasses/UserController.java]: Same problem as in ChangePasswordValidator.java

**Comment by santos-samuel** [comment]
Is there any folder which is ignored by the compiler where i can put these files as external files and not as actual project files?
Because removing those dependencies implies changing most part of the method which will cause de bug to dissapear.

In other words, is it possible not to treat the `spoon.getExecutableDeclarationBugTest.testClasses` as a package?

**Comment by monperrus** [comment]
> Is there any folder which is ignored by the compiler

yes, `src/test/resources`

**Comment by monperrus** [comment]
cloned locally, had a look.

`getExecutableDeclaration` returns null because the argument types do not match:

* in `UserController` you call `validate(pt.ist.socialsoftware.edition.ldod.forms.ChangePasswordForm, org.springframework.validation.BindingResult)`
* in `ChangePasswordValidator`, validate receives as argument `validate(Object, org.springframework.validation.Errors)` 

org.springframework.validation.BindingResult != org.springframework.validation.Errors

**Comment by santos-samuel** [comment]
Oh i see... So it isn't a bug... Spoon only doesn't know that BindingResult is hierarchically related with Errors.
Ok, that closes this issue. Thank you for your availability!

**Comment by monperrus** [comment]
thanks for the super useful failing test case!

**Pull request closed without merging**